docs: update installation instructions for CAM with FTorch (#19)
DataWaveProject/CAM: README.md

```
git clone https://github.com/DataWaveProject/CAM.git
cd CAM

> git checkout coupling

Reviewer: I would be a little careful here.
`coupling` is the current working branch with the latest code, but it will eventually be merged back into `datawave_ml` as the main stable branch for this project.

Author: Ah, that makes sense. I wrote this based on what I actually did. Currently, if you run with `datawave_ml` it won't work. But I can see why you want to keep it. Do you think we could merge `coupling` into `datawave_ml` soon? or will it be the last thing?

Maybe we could add a note that for the time being users will need to use `coupling`?

Reviewer: I'm hoping so now that the issues with NaNs are resolved.
I will bring it up at today's meeting.

A note saying to use coupling for now - or perhaps use "your relevant development branch"? - would work, and I can make it a task on #14 to update this in docs before merging.

Author: let's wait until tomorrow. If `coupling` is being merged soon then I am happy to leave it as is. Otherwise I will add more text

Reviewer: OK, Qiang will try building this week (he says) so let's merge with `coupling` for now, I've added a note to my coupling PR to update to `datawave_ml` before that gets merged.

```
This branch is built upon the `cam6_3_139` tag from the
[main ESMCOMP/CAM repository](https://github.com/ESCOMP/CAM).

### Obtaining `FTorch`

> To use PyTorch-based neural nets in CAM, we first need to build and link `FTorch`.

Reviewer: Nit: We only build at this stage, linking occurs during build once we have FTorch compatible CIME which is now described in a separate section below.
```suggestion
To use PyTorch-based neural nets in CAM, we use [`FTorch`](https://github.com/Cambridge-ICCS/FTorch) which needs to be built on the system before we build CAM.
```

> To install `FTorch` on Derecho follow the instructions in section [`FTorch` on Derecho](#ftorch-on-derecho) below.

Reviewer 2: Do we want to add a placeholder section for building on CSD3 here now, or do we cross that bridge when we come to it?

Author: Good idea, but I don't think we should add them until we have the instructions ready to go.

We should also consider/discuss who these instructions are for. My understanding is it's for users who will most likely only run on `derecho` or other NCAR systems. So it might add confusion.

Either way, I do agree that we should write up `CSD3` instructions, but I am not sure yet, where they would best belong -- perhaps we can discuss in our next internal meeting.

Reviewer: I agree.
I am hoping that I will have time in January to work on FTorch-CESM-CSD3 and make this overall process much simpler, but for now we should leave it until we have something to write about. To my mind documentation to run on CSD3 would go into the CSD3 documentation pages alongside more general docs about running CESM on CSD3.

I would also second Tom's point that this is intended as a resource specifically for our DataWave partners who run on Derecho at this point. We should leave it as specific to them, and adapt as necessary should they need to run on other systems.

You can then navigate to the case directory at `<path_to_testcase_directory>`.

### Build CAM with `FTorch`

Before we can run `./case.build` we first need to make some manual changes to the `Makefile` located in

> `path_to_testcase_directory>/Tools/Makefile`. This will allow the CIME build system to locate `FTorch`.

Reviewer: ```suggestion
`<path_to_testcase_directory>/Tools/Makefile`. This will allow the CIME build system to locate `FTorch`.
```

From the test case directory, modify `Tools/Makefile` line 602 to set the environment variable `FTORCH_LIB` to the location of
the `FTorch` library on your system.

If you followed the instructions in section [`FTorch` on Derecho](#ftorch-on-derecho) this will be `$HOME/FTorch/bin/ftorch_intel`.

```make
FTORCH_LIB := $HOME/FTorch/bin/ftorch_intel
```

### Setting up case details

We can now run `./case.setup` from within the case directory. Once this has been done then edit the generated `user_nl_cam` in
the case directory as required. Add the following lines:

```fortran
gw_convect_dp_ml=.false.
gw_convect_dp_ml_compare=.true.
gw_convect_dp_ml_net_path='/path/to/neural/net'
gw_convect_dp_ml_norms='/path/to/norms'

> ```

Reviewer: This is an example configuration, but users can vary for their applications.
Adding the explanations below is really helpful, thanks @TomMelt but maybe shoudl add a note to say that this config is an example and can be tailored to your particular experiment as per the descriptions below?

Author: I have added some more wording [here](https://github.com/DataWaveProject/CAM/blob/f8c4ae9d6466d2631e998396ba86e6944f2838d7/README.md). lmk what you think

Reviewer: Yeah, seems good.

* `gw_convect_dp_ml` (`logical`)

   Whether or not to use the ML scheme for gravity waves produced by deep convection. Default: `.false.`

* `gw_convect_dp_ml_compare` (`logical`)

   Whether or not to run a piggybacking comparison of the ML deep convection gravity waves to the original scheme. Only one
   scheme will be used to advance the simulation as dictated by `gw_convect_deep_ml`. Default: `.false.`

* `gw_convect_dp_ml_net_path`

   Absolute filepath to the deep convection gravity wave neural net used when `gw_convect_dp_ml` is set to `.true.` (`.pt`
   extension).

* `gw_convect_dp_ml_norms`

   Absolute filepath to the deep convection gravity wave normalisation weights (NetCDF) used when `gw_convect_dp_ml` is set to `.true.`.

> [!TIP]
> Consider adding the following to generate output diagnostics of variables as desired.
>
> ```fortran
> fincl<n> = 'MYVAR'
> ```

We can then run `./case.build` from within the case directory to build the model.

The case can be run with `./case.submit` from the case directory.

> [!NOTE]

> > By default CESM will place output in `$SCRATCH/case/` and logs/restart files in `$SCRATCH/archive/case/`.

Reviewer: ```suggestion
> By default CESM will place outputs in `$SCRATCH/case/` essential parts of which will be moved to `$SCRATCH/archive/case/` after run completion.
```

> > To place all output with logs in `archive/case` switch 'short term archiving' on by running `./xmlchange DOUT_S=FALSE` in the

Reviewer: ```suggestion
> To leave all output in `$SCRATCH/case/` switch 'short term archiving' off by running `./xmlchange DOUT_S=FALSE` in the
```

Please see the [wiki](https://github.com/ESCOMP/CAM/wiki) for complete documentation on CAM, getting started with git and how to contribute to CAM's development.

### `FTorch` on Derecho

The following steps can be followed to ensure `FTorch` is built to be consistent
with CAM on Derecho.

#### load CAM environment

> For compatibility with CAM we need to be specific about the environment and compilers we load. The following sequence of modules

Reviewer: ```suggestion
For compatibility with the version of CAM we are using (branched from the `cam6_3_139` tag) we need to be specific about the environment and compilers we load. The following sequence of modules
```

are required to build `FTorch` compatible with the intel build of CAM on Derecho:

```bash
module purge
module load ncarenv/23.06
module load intel-oneapi/2023.0.0
module load mkl
module load cmake
module load cuda/11.7.1
```

> [!NOTE]
> In future builds or releases, or on different machines, the environment for building CAM may change. In this case the `FTorch`
> environment should be updated accordingly.

> [!NOTE]
> When building `FTorch` with `cmake` in the next step, we use the absolute path to `libtorch`. Previously, we loaded the
> `libtorch` module i.e., `module load libtorch/2.1.2` but this conflicts with `ncarenv/23.06`. We do not actually need to load
> the `libtorch` module to use the absolute path, so for now we can ignore this.

> 

Reviewer: I wonder if it is just easier to not mention this to avoid potential confusion?
"Previously" isn't really relevant here as to my knowledge no-one from NCAR/DataWave has yet followed these instructions.

Author: Happy to take it out

Author: removed in [269a5d0](https://github.com/DataWaveProject/CAM/pull/19/commits/269a5d0b592b1285b415601819a42795128ade76)
